enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java

```diff
@@ -791,10 +796,75 @@ private void markAsWaitingFor( ForsetiLockManager.Lock lock, ResourceType type,
             // after we've generated a description of it.
             if ( lock.detectDeadlock( id() ) != -1 )
             {
-                waitList.clear();
-                throw new DeadlockDetectedException( message );
+                // If the deadlock is real, then an owner of this lock must be (transitively) waiting on a lock that
+                // we own. So to verify the deadlock, we traverse the lock owners and their `waitingForLock` fields,
+                // to find a lock that has us among the owners.
+                // We only act upon the result of this method if the `tries` count is above some threshold. The reason
+                // is that the Lock.collectOwners, which is algorithm relies upon, is inherently racy, and so only
+                // reduces the probably of a false positive, but does not eliminate them.
+                if ( isDeadlockReal( lock, tries ) )
+                {
+                    // After checking several times, this really does look like a real deadlock.
+                    waitList.clear();
+                    waitingForLock = null;
+                    throw new DeadlockDetectedException( message );
+                }
             }
         }
+        waitingForLock = null;
+    }
+
+    private boolean isDeadlockReal( ForsetiLockManager.Lock lock, int tries )
+    {
+        Set<ForsetiLockManager.Lock> waitedUpon = new HashSet<>();
```

Comment:
@jakewins that's what I've been saying all along 😉  - which is a poor way of saying that I agree.

Actually, I don't mind if we use the dreadlocks algorithm for initial deadlock detection or not, and then fall back to complete resource graph traversal as in this patch, if that speeds things up while not complicating the code much and without having a big cost overhead, I think it's a great idea. But I agree that the focus should be on making uncontended acquisition fast, whatever we do before going to sleep waiting for a lock is minimal compared to the cost overhead of actually waiting. I also, firmly, believe that we should never ever report a false deadlock to users.